### PR TITLE
Python client cancel jobs

### PR DESCRIPTION
**Pull request by freddyaboulton**
# Description

Makes jobs cancellable. In general we can't tell the server that it should stop running the job without somehow adding a route to cancel running tasks but that will not exist for older apps and also that could get spammed/exploited.

So this is the equivalent of closing the Chrome tab running the UI. The server will stop running the job the next time it tries to send a message to the client. This is still useful as this means jobs will be dequeued and iterative jobs will finish their iterations early!

# Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have added a short summary of my change to the CHANGELOG.md
- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

# A note about the CHANGELOG

Hello 👋 and thank you for contributing to Gradio!

All pull requests must update the change log located in CHANGELOG.md, unless the pull request is labeled with the "no-changelog-update" label.

Please add a brief summary of the change to the Upcoming Release > Full Changelog section of the CHANGELOG.md file and include
a link to the PR (formatted in markdown) and a link to your github profile (if you like). For example, "* Added a cool new feature by `[@myusername](link-to-your-github-profile)` in `[PR 11111](https://github.com/gradio-app/gradio/pull/11111)`".

If you would like to elaborate on your change further, feel free to include a longer explanation in the other sections.
If you would like an image/gif/video showcasing your feature, it may be best to edit the CHANGELOG file using the 
GitHub web UI since that lets you upload files directly via drag-and-drop.

### CODE REVIEW

File: client/python/gradio_client/client.py [hunk 1]
```diff
@@ -797,3 +810,24 @@ def status(self) -> StatusUpdate:
     def __getattr__(self, name):
         """Forwards any properties to the Future class."""
         return getattr(self.future, name)
+
+    def cancel(self) -> bool:
```
Comment (abidlabs) [inline client/python/gradio_client/client.py:814]: Awesome docstring!

File: client/python/gradio_client/client.py [hunk 2]
```diff
@@ -755,13 +755,26 @@ def status(self) -> StatusUpdate:
...
+                eta=None,
+            )
         if self.done():
             if not self.future._exception:  # type: ignore
                 return StatusUpdate(
                     code=Status.FINISHED,
                     rank=0,
                     queue_size=None,
-                    success=True,
+                    success=False,
```
Comment (abidlabs) [inline client/python/gradio_client/client.py:777]: Why is this False?

Comment (freddyaboulton) [reply client/python/gradio_client/client.py:777]: Lol typo - will revert!

**Comment by abidlabs** [comment]
Thanks @freddyaboulton, tested this out quite a bit. Generally, it works super well! The only issue I was able to find is that if a job is cancelled _while it is iterating_, and then a subsequent job is submitted, then it will pick up from where the previous iterator left off, rather than starting from the beginning. I don't think this is expected behavior and doesn't match the behavior when you stop and rerun a generator through the UI.

Example:

```py
from gradio_client import Client
import time

client = Client("abidlabs/test-yield")
job1 = client.submit("abcdef")
time.sleep(3)
job1.cancel()

print("Job 1")
for output in job1.outputs():
    print(output)

job2 = client.submit("abcdef")

print("\nJob 2")
for output in job2:
    print(output)
```

**Comment by abidlabs** [comment]
LGTM @freddyaboulton, thanks for great PR!

**Comment by abidlabs** [comment]
I'll go ahead and merge this in

**Comment by freddyaboulton** [comment]
Thanks for the review @abidlabs !

**Pull request merged**